src/portfft/descriptor.hpp

```diff
@@ -367,7 +367,12 @@ class committed_descriptor {
     }
     std::vector<std::tuple<detail::level, std::vector<sycl::kernel_id>, std::vector<Idx>>> param_vec;
     auto check_and_select_target_level = [&](IdxGlobal factor_size, bool batch_interleaved_layout = true) -> bool {
-      if (detail::fits_in_wi<Scalar>(factor_size)) {
+      constexpr std::size_t LocalRange = PORTFFT_SGS_IN_WG * SubgroupSize;
+      if (detail::fits_in_wi<Scalar>(factor_size) &&
+          static_cast<std::size_t>(local_memory_size) >
+              sizeof(Scalar) * 2 * LocalRange * static_cast<std::size_t>(factor_size)) {
+        // The local memory requirement (LocalRange * factor_size) is only required for the final factor.
+        // There is no way to know if this is the last factor, so it is always used.
```

Comment:
The `kernel_data_struct` has a field for `local_mem_required`. Can't we use this as a proper solution to check that each kernel has enough local memory?

Reply:
I don't think so:
* `local_mem_required` represents the input/output mem requirements in the global impl - see [here](https://github.com/codeplaysoftware/portFFT/blob/c28df725892aabc60054d914fc55c43b4b352417/src/portfft/common/global.hpp#L470)
* It is currently computed after the implementation is chosen [here](https://github.com/codeplaysoftware/portFFT/blob/c28df725892aabc60054d914fc55c43b4b352417/src/portfft/dispatcher/global_dispatcher.hpp#L108). I think we need a larger refactor to avoid circular dependencies.

@AD2605 I suspect understands the system better than I do?

Reply:
That's correct. The `local_mem_required` would not be populated at this point, as we are still in `prepare_implementation`, for that matter the `kernel_data_struct` would not have been created. We would always need to calculate the local memory before obtaining the set of `kernel_ids` for that factor, as we would need to ascertain that we do not run out of it.

Reply:
Alright, thanks for the explanations!